app/mci/pkg/gcp/loadbalancer/loadbalancersyncer.go

```diff
@@ -46,32 +49,115 @@ const (
 	// Prefix used by the namer to generate names.
 	// This is used to identify resources created by this code.
 	mciPrefix = "mci1"
+
+	// instanceGroupsAnnotationKey is the annotation key used by gclb ingress
+	// controller to specify the name and zone of instance groups created
+	// for the ingress.
+	// TODO(nikhiljindal): Refactor the code in kubernetes/ingress-gce to
+	// export it and then reuse it here.
+	instanceGroupsAnnotationKey = "ingress.gcp.kubernetes.io/instance-groups"
 )
 
 type LoadBalancerSyncer struct {
 	lbName string
 	hcs    healthcheck.HealthCheckSyncerInterface
+	bss    backendservice.BackendServiceSyncerInterface
 	client kubeclient.Interface
+	igp    ingressig.InstanceGroups
 }
 
 func NewLoadBalancerSyncer(lbName string, client kubeclient.Interface, cloud *gce.GCECloud) *LoadBalancerSyncer {
 	namer := utilsnamer.NewNamer(mciPrefix, lbName)
 	return &LoadBalancerSyncer{
 		lbName: lbName,
 		hcs:    healthcheck.NewHealthCheckSyncer(namer, cloud),
+		bss:    backendservice.NewBackendServiceSyncer(namer, cloud),
 		client: client,
+		igp:    cloud,
 	}
 }
 
 func (l *LoadBalancerSyncer) CreateLoadBalancer(ing *v1beta1.Ingress) error {
+	var err error
 	ports := l.ingToNodePorts(ing)
-	// Create health check to be used by the backend service.
-	if err := l.hcs.EnsureHealthCheck(l.lbName, ports); err != nil {
+	// Create health checks to be used by the backend service.
+	healthChecks, hcErr := l.hcs.EnsureHealthCheck(l.lbName, ports)
+	if hcErr != nil {
+		// Keep aggregating errors and return all at the end, rather than giving up on the first error.
+		err = multierror.Append(err, hcErr)
+	}
+	igs, namedPorts, err := l.getIGsAndNamedPorts(ing)
+	// Cant really create any backend service without named ports. No point continuing.
+	if err != nil {
+		return err
+	}
+	// Create backend service. This should always be called after the health check since backend service needs to point to the health check.
+	if err := l.bss.EnsureBackendService(l.lbName, ports, healthChecks, namedPorts, igs); err != nil {
 		return err
 	}
 	return nil
 }
 
+// Returns links to all instance groups and named ports on any one of them.
+// Note that it picks an instance group at random and returns the named ports for that instance group, assuming that the named ports are same on all instance groups.
+// Also note that this returns all named ports on the instance group and not just the ones relevant to the given ingress.
+func (l *LoadBalancerSyncer) getIGsAndNamedPorts(ing *v1beta1.Ingress) ([]string, backendservice.NamedPortsMap, error) {
+	// Keep trying until ingress gets the instance group annotation.
+	for try := true; try; time.Sleep(1 * time.Second) {
```

Comment:
Should we worry about having an exponential backoff?

Reply:
I dont think so (could be convinced otherwise).

User will be waiting and we need to keep printing the trying message at constant rate to ensure that the tool is working and is not dead.

Reply:
I don't think we need to do anything now- add a TODO that we might need backoff here?

If we want to keep the user updated, we can always print how long we're waiting before trying again. There shouldn't be a ton of these outstanding requests to the API server such that it causes a problem.

Reply:
Added the TODO.
Note that there will always be only one outstanding request. We wait for the request to complete and return before trying again.